src/XMonad/StackSet.hs

```diff
@@ -178,6 +180,23 @@ data Stack a = Stack { focus  :: !a        -- focused thing in this set
     deriving (Show, Read, Eq)
 
 
+instance Functor Stack where
+    fmap f s = Stack
+        (f (focus s))
+        (fmap f (up s))
+        (fmap f (down s))
+
+instance Foldable Stack where
+    toList = integrate
+    foldr f z = foldr f z . toList
+
+instance Traversable Stack where
+    traverse f s =
+        flip Stack
+            <$> fmap getReverse (traverse f (Reverse (up s)))
+            <*> f (focus s)
+            <*> traverse f (down s)
```

Comment:
I can see that these are law abiding, but it would perhaps be nice to also tell this to `QuickCheck` :)

Reply:
I don't know QuickCheck well enough to test ``forall (t :: Applicative n => m a -> n a)  f. t . traverse f = traverse (t . f)`` or ``forall g f. traverse (Compose . fmap g . f) = Compose . fmap (traverse g) . traverse f``.

I could write a test for ``traverse Identity = Identity``, which would be telling.

Reply:
You're right, this is probably not as straightforward as I thought it would be.  

The easiest way, I suppose, would be to depend on [some library](https://hackage.haskell.org/package/quickcheck-classes-0.6.4.0/docs/Test-QuickCheck-Classes.html#v:traversableLaws) that has already solved this problem; I don't think this would be much of an issue, as it'd be solely in the test-suite.

Then again, we're really only sticking together known 'Foldable' and 'Traversable's, so I'm not sure if this is worth it after all.  @liskin @byorgey @psibi opinions?

Reply:
I wouldn't insist on having the instance tested, personally, but that's proably because I don't remember being bitten by a non-law-abiding instance. The coding style section of XMonad.Doc.Developing states that all pure functions must have quickcheck properties, so if one can be added relatively easily, it'd be nice. `quickcheck-classes` isn't in lts-12, however, so it may be somewhat painful (probably less painful than pandoc, though, so probably tolerable? I have no idea any more).

Reply:
>I wouldn't insist on having the instance tested, personally, but that's proably because I don't remember being bitten by a non-law-abiding instance. 

I have to confess that I was at one point; maybe that's why I'm so scared about this :) 

I do, however, think that if we _do_ decide to test for the properties in question, we should test for _all_ of them, and not only a few selected ones (probably with the help of some library)

Reply:
So should I just remove the identity test? I understand the importance of testing, but for something like this, honestly, testing is _not good enough_. It should be _provably_ correct.

Reply:
I mean, sure I can prove that the instances are correct, but short of using something like liquid haskell (can it do this even?) I can't really convince GHC of this...  I think you should remove the identity test in every case though, yes.

Two ways forward that I can see

  1. Test _everything_ with the use of some library (additional dependencies _shouldn't_ matter because they would only be compiled on a `stack test` or equivalent)
  2. Test nothing because the instances are "obvious" (they are, but so are a few of the pure functions and yet they are still quickchecked, as `XMonad.Doc.Extending` demands)

@geekosaur do you have an opinion on this?

Reply:
In the past, verification has been done with proof assistants rather than Haskell-based testing. This, for example, is why the `StackSet` takes so many type variables when they're only ever instantiated to fixed types: they were instantiated differently in (iirc) Coq since the X11 types weren't available. That doesn't mean we're forced to go the same way now, but it is a data point.

(I'm not sure we even have the scaffolding for the old tests, which could probably stand to be re-run. Someone would have to ask @donsbot what happened to them, since I don't think they're in the repo.)

In the meantime, simple testing is probably worth doing but it's not worth trying to get too fancy; if we can't reasonably test something in Haskell then it's not worth bothering.

Reply:
Alright, @wygulmage I think it's then worth going for option 1. and use a library to fully quickcheck these instances.

Reply:
OK. I have updated the tests to use quickcheck-classes (and removed the 'ad-hoc' `Identity` test). This was much easier than I expected. I had to add an `Arbitrary` instance for `Stack`, which I am slightly less confident in than the `Traversable` instance.

To make the tests compile with GHC 8.4.4, I had to add `Eq1` and `Show1` instances for `Stack`. I am _much_ less confident in the `Show1` instance, but I don't know if it can be properly derived. I'm not sure if it matters, because nothing else depends on it. I could use a newtype for testing and define the instances on that to avoid exporting them for `Stack`.

Reply:
> To make the tests compile with GHC 8.4.4, I had to add `Eq1` and `Show1` instances for `Stack`. I am _much_ less confident in the `Show1` instance, but I don't know if it can be properly derived. I'm not sure if it matters, because nothing else depends on it. I could use a newtype for testing and define the instances on that to avoid exporting them for `Stack`.

I think that we should either move these into `test` (they would become orphans at that point, but since it's a testing repository this may or may not matter) or use the afforementioned newtype

Reply:
Orphans are acceptable in executables, including tests.

Reply:
I've switched the tests and `Arbitrary`, `Show1`, and `Eq1` instances to a 'ShowStack` newtype in Properties/Stack.hs.